node/impl/full/eth.go

```diff
@@ -2135,3 +2141,93 @@ func (g gasRewardSorter) Swap(i, j int) {
 func (g gasRewardSorter) Less(i, j int) bool {
 	return g[i].premium.Int.Cmp(g[j].premium.Int) == -1
 }
+
+func (a *EthModule) EthGetBlockReceipts(ctx context.Context, blockParam ethtypes.EthBlockNumberOrHash) ([]*api.EthTxReceipt, error) {
+	ts, err := getTipsetByEthBlockNumberOrHash(ctx, a.Chain, blockParam)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get tipset: %w", err)
+	}
+
+	// Execute the tipset to get the receipts, messages, and events
+	st, msgs, receipts, err := executeTipset(ctx, ts, a.Chain, a.StateAPI)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to execute tipset: %w", err)
+	}
+
+	// Load the state tree
+	stateTree, err := a.StateManager.StateTree(st)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to load state tree: %w", err)
+	}
+
+	// Get the parent tipset CID
+	parentTsCid, err := ts.Parents().Cid()
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get parent tipset cid: %w", err)
+	}
+
+	// Load executed messages and extract events
+	executedMsgs, err := a.EthEventHandler.EventFilterManager.LoadExecutedMessages(ctx, ts, ts)
```

Comment:
ah, we have a problem here - you can't pass in `ts` to this, you need `ts+1` which you don't have because you synthetically generated the next state, but if you look inside `LoadExecutedMessages` you'll see that these two parameters are only needed to gather `msgs` and reference `ParentMessageReceipts`, which you have as `receipts`. So what you probably want to do is make a new function internal to `EventFilterManager` that wraps `LoadExecutedMessages` and make it take msgs and receipts as arguments. You could even detach it from `EventFilterManager` entirely so you don't need to reach into that dependency (it also makes this independent of the events db, which is an interesting property).

Here's a quick go at what that might look like:

```diff
diff --git a/chain/events/filter/event.go b/chain/events/filter/event.go
index fa17d235e..2ce31e127 100644
--- a/chain/events/filter/event.go
+++ b/chain/events/filter/event.go
@@ -16,6 +16,7 @@ import (
        "github.com/filecoin-project/go-state-types/abi"
        blockadt "github.com/filecoin-project/specs-actors/actors/util/adt"
 
+       "github.com/filecoin-project/lotus/chain/actors/adt"
        cstore "github.com/filecoin-project/lotus/chain/store"
        "github.com/filecoin-project/lotus/chain/types"
 )
@@ -438,37 +439,43 @@ func (m *EventFilterManager) loadExecutedMessages(ctx context.Context, msgTs, rc
                return nil, xerrors.Errorf("read messages: %w", err)
        }
 
-       st := m.ChainStore.ActorStore(ctx)
+       store := m.ChainStore.ActorStore(ctx)
 
-       arr, err := blockadt.AsArray(st, rctTs.Blocks()[0].ParentMessageReceipts)
+       arr, err := blockadt.AsArray(store, rctTs.Blocks()[0].ParentMessageReceipts)
        if err != nil {
                return nil, xerrors.Errorf("load receipts amt: %w", err)
        }
 
-       if uint64(len(msgs)) != arr.Length() {
-               return nil, xerrors.Errorf("mismatching message and receipt counts (%d msgs, %d rcts)", len(msgs), arr.Length())
-       }
-
-       ems := make([]executedMessage, len(msgs))
-
-       for i := 0; i < len(msgs); i++ {
-               ems[i].msg = msgs[i]
-
-               var rct types.MessageReceipt
-               found, err := arr.Get(uint64(i), &rct)
+       receipts := make([]types.MessageReceipt, arr.Length())
+       for i := 0; i < len(receipts); i++ {
+               found, err := arr.Get(uint64(i), &receipts[i])
                if err != nil {
                        return nil, xerrors.Errorf("load receipt: %w", err)
                }
                if !found {
                        return nil, xerrors.Errorf("receipt %d not found", i)
                }
-               ems[i].rct = &rct
+       }
+
+       return LoadExecutedMessages(ctx, store, msgs, receipts)
+}
+
+func LoadExecutedMessages(ctx context.Context, store adt.Store, msgs []types.ChainMsg, receipts []types.MessageReceipt) ([]executedMessage, error) {
+       if len(msgs) != len(receipts) {
+               return nil, xerrors.Errorf("mismatching message and receipt counts (%d msgs, %d rcts)", len(msgs), len(receipts))
+       }
+
+       ems := make([]executedMessage, len(msgs))
 
-               if rct.EventsRoot == nil {
+       for i := 0; i < len(msgs); i++ {
+               ems[i].msg = msgs[i]
+               ems[i].rct = &receipts[i]
+
+               if receipts[i].EventsRoot == nil {
                        continue
                }
 
-               evtArr, err := amt4.LoadAMT(ctx, st, *rct.EventsRoot, amt4.UseTreeBitWidth(types.EventAMTBitwidth))
+               evtArr, err := amt4.LoadAMT(ctx, store, *receipts[i].EventsRoot, amt4.UseTreeBitWidth(types.EventAMTBitwidth))
                if err != nil {
                        return nil, xerrors.Errorf("load events amt: %w", err)
                }
@@ -491,7 +498,6 @@ func (m *EventFilterManager) loadExecutedMessages(ctx context.Context, msgTs, rc
                if err != nil {
                        return nil, xerrors.Errorf("read events: %w", err)
                }
-
        }
 
        return ems, nil
```

It also gets us close, but not quite, to being able to remove some duplicate code for loading events which you'll also see in node/impl/full/chain.go `ChainGetEvents`. But that can be a job for some other time.